controllers/maintenance_mode_checker.go

```diff
@@ -0,0 +1,121 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	fdbv1beta2 "github.com/FoundationDB/fdb-kubernetes-operator/api/v1beta2"
+)
+
+// updatePods provides a reconciliation step for recreating pods with new pod
+// specs.
+type maintenanceModeChecker struct{}
+
+const (
+	podUpCheckDelayInSeconds = 60
+)
+
+// reconcile runs the reconciler's work.
+func (maintenanceModeChecker) reconcile(ctx context.Context, r *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster) *requeue {
+	logger := log.WithValues("namespace", cluster.Namespace, "cluster", cluster.Name, "reconciler", "maintenanceModeChecker")
+
+	adminClient, err := r.getDatabaseClientProvider().GetAdminClient(cluster, r.Client)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	defer adminClient.Close()
+
+	maintenanceZone, err := adminClient.GetMaintenanceZone()
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	// TODO
+	logger.Info("maintenance mode zone", "zone", maintenanceZone)
+	// Cluster is not in maintenance mode
+	if maintenanceZone == "" {
+		if cluster.Status.MaintenanceModeInfo.ZoneId != "" {
+			/// TODO
+			logger.Info("Clearing maintenance mode HEREREE")
+			cluster.Status.MaintenanceModeInfo = fdbv1beta2.MaintenanceModeInfo{}
+			err = r.Status().Update(ctx, cluster)
+			if err != nil {
+				return &requeue{curError: err}
+			}
+		}
+		// TODO
+		logger.Info("HHHHHHHHH")
+		return nil
+	}
+	// FDB Cluster is in maintenance mode but not due to this operator actions
+	if maintenanceZone != cluster.Status.MaintenanceModeInfo.ZoneId {
+		if cluster.Status.MaintenanceModeInfo.ZoneId != "" {
+			/// TODO
+			logger.Info("Clearing maintenance mode TTTTTTTHEREREE")
+			cluster.Status.MaintenanceModeInfo = fdbv1beta2.MaintenanceModeInfo{}
+			err = r.Status().Update(ctx, cluster)
+			if err != nil {
+				return &requeue{curError: err}
+			}
+		}
+		// TODO
+		logger.Info("DDDDDDDDDDD")
+		return nil
+	}
+	logger.Info("Cluster in maintenance mode", "zone", maintenanceZone)
+	// FDB Cluster is in maintenance mode due to this operator actions
+	if cluster.Status.MaintenanceModeInfo.StartTimestamp.Add(time.Duration(podUpCheckDelayInSeconds) * time.Second).After(time.Now()) {
+		return &requeue{message: "Waiting for delay to expire for checking if pods under maintenance mode have been bounced", delayedRequeue: true}
+	}
+	status, err := adminClient.GetStatus()
+	if err != nil {
+		return &requeue{curError: err}
+	}
+	processGroupsToCheck := make(map[string]struct{})
+	for _, id := range cluster.Status.MaintenanceModeInfo.ProcessGroups {
```

Comment:
This has a race condition and we probably want to check the uptime of the processes otherwise we could run into the issue that we set the maintenance mode, the operator triggers the deletion of the Pods, which might be delayed by Kubernetes and the Pods which are still in deletion still have the processes running and now we would remove the maintenance mode before those Pods are actually recreated. Which would result in data distribution.

Reply:
+1

Reply:
I am using a 60 second delay as in this variable podUpCheckDelayInSeconds and then check if the pod is present in the status to account for the race condition. I agree that using uptime is a better approach.